### PR TITLE
Add options validation to collection

### PR DESCRIPTION
**Pull request by kvwalker**
Fixes NODE-1671

### CODE REVIEW

File: lib/collection.js [hunk 1]
```diff
@@ -737,23 +769,37 @@ Collection.prototype.insert = deprecate(function(docs, options, callback) {
...
-  if (this.s.options.ignoreUndefined) {
-    options = Object.assign({}, options);
-    options.ignoreUndefined = this.s.options.ignoreUndefined;
-  }
+  options = validate(
+    updateOneSchema,
+    options,
+    {},
+    {
+      ignoreUndefined: this.s.options.ignoreUndefined,
```
Comment (kvwalker) [inline lib/collection.js]: In `validate`, we should be checking that overrides/dynamic defaults being applied have values (i.e. `!= null`).

File: lib/collection.js [hunk 2]
```diff
@@ -1559,14 +1858,33 @@ Collection.prototype.stats = function(options, callback) {
  * @param {Collection~findAndModifyCallback} [callback] The collection result callback
  * @return {Promise<Collection~findAndModifyWriteOpResultObject>} returns Promise if no callback passed
  */
+const findOneAndDeleteSchema = {
+  collation: { type: 'object' },
+  projection: { type: 'object' },
+  sort: { type: 'object' },
+  maxTimeMS: { type: 'number' },
+  session: { type: ClientSession },
+  fields: { overrideOnly: true },
```
Comment (kvwalker) [inline lib/collection.js]: make a ticket for `fields` to be internally renamed to `projection` (or something).

File: lib/collection.js [hunk 3]
```diff
@@ -1782,8 +2168,20 @@ Collection.prototype.aggregate = function(pipeline, options, callback) {
   // Build the command
   const command = { aggregate: this.s.name, pipeline: pipeline };
 
+  const finalOptions = validate(
```
Comment (kvwalker) [inline lib/collection.js]: `finalOptions` => `options`

File: lib/collection.js [hunk 4]
```diff
@@ -1809,64 +2207,55 @@ Collection.prototype.aggregate = function(pipeline, options, callback) {
...
   // Do we have a readConcern specified
   if (!ignoreReadConcern) {
-    decorateWithReadConcern(command, this, options);
+    decorateWithReadConcern(command, this, finalOptions);
   }
 
   // If we have allowDiskUse defined
-  if (options.allowDiskUse) command.allowDiskUse = options.allowDiskUse;
-  if (typeof options.maxTimeMS === 'number') command.maxTimeMS = options.maxTimeMS;
+  if (finalOptions.allowDiskUse) command.allowDiskUse = finalOptions.allowDiskUse;
```
Comment (kvwalker) [inline lib/collection.js]: make a ticket for this to use `decorateCommand` (maxTimeMS, hint, explain, allowDiskUse)

File: lib/collection.js [hunk 5]
```diff
@@ -1809,64 +2207,55 @@ Collection.prototype.aggregate = function(pipeline, options, callback) {
...
-  options.cursorFactory = AggregationCursor;
   if (typeof callback !== 'function') {
     if (!this.s.topology.capabilities()) {
       throw new MongoError('cannot connect to server');
     }
 
     // Allow disk usage command
-    if (typeof options.allowDiskUse === 'boolean') command.allowDiskUse = options.allowDiskUse;
-    if (typeof options.maxTimeMS === 'number') command.maxTimeMS = options.maxTimeMS;
+    if (typeof finalOptions.allowDiskUse === 'boolean') {
```
Comment (kvwalker) [inline lib/collection.js]: add to ticket

File: lib/options_validator.js [hunk 6]
```diff
@@ -71,68 +94,92 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
...
   const verifiedOptions = Object.assign({}, providedOptions);
 
-  if (validationLevel === VALIDATION_LEVEL_NONE && overrideOptions == null) {
-    return Object.freeze(verifiedOptions);
+  if (
+    validationLevel === VALIDATION_LEVEL_NONE &&
+    overrideOptions == null &&
+    optionDefaults == null
+  ) {
+    return verifiedOptions;
```
Comment (kvwalker) [inline lib/options_validator.js]: This won't set non-dynamic defaults.

File: lib/options_validator.js [hunk 7]
```diff
@@ -71,68 +94,92 @@ function validate(optionsSchema, providedOptions, overrideOptions, validationOpt
...
   });
 
-  return Object.freeze(verifiedOptions);
+  return verifiedOptions;
+}
+
+function checkType(requiredType, providedValue) {
+  if (typeof requiredType === 'string') {
+    if (
+      (requiredType === 'object' && isObject(providedValue)) ||
```
Comment (kvwalker) [inline lib/options_validator.js]: `if(requiredType === 'obect') return isObject(providedValue);`

File: test/functional/find_tests.js [hunk 8]
```diff
@@ -2737,14 +2737,11 @@ describe('Find', function() {
...
         var options = { raw: 'TEST' };
-        collection.find({}, options, function(error) {
-          test.equal(null, error);
-          test.equal(undefined, options.skip);
-          test.equal(undefined, options.limit);
-          test.equal('TEST', options.raw);
-          client.close();
-          done();
-        });
+        expect(() => collection.find({}, options)).to.throw(
```
Comment (kvwalker) [inline test/functional/find_tests.js]: delete

File: test/functional/readpreference_tests.js [hunk 9]
```diff
@@ -246,7 +246,7 @@ describe('ReadPreference', function() {
           };
 
           // Perform the map reduce
-          collection.mapReduce(map, reduce, { out: 'inline' }, function(/* err */) {
+          collection.mapReduce(map, reduce, { out: { inline: 1 } }, function(/* err */) {
```
Comment (kvwalker) [inline test/functional/readpreference_tests.js]: out can be a string or document. fix in `collection.js`

File: lib/collection.js [hunk 10]
```diff
@@ -65,6 +64,8 @@ const updateDocuments = require('./operations/collection_ops').updateDocuments;
 const updateMany = require('./operations/collection_ops').updateMany;
 const updateOne = require('./operations/collection_ops').updateOne;
 
+const DEFAULT_VALIDATION = 'error';
```
Comment (kvwalker) [inline lib/collection.js]: I'm temporarily setting this to `error` so we can catch any test failures. I'll change it back before merging.

File: lib/options_validator.js [hunk 11]
```diff
@@ -106,11 +106,10 @@ function validate(
...
     }
 
-    if (
-      optionDefaults &&
-      optionDefaults.hasOwnProperty(optionName) &&
-      optionDefaults[optionName] != null
-    ) {
+    if (optionDefaults && optionDefaults.hasOwnProperty(optionName)) {
+      if (optionDefaults[optionName] == null) {
+        throw new Error(`A dynamic default of null or undefined was provided for [${optionName}].`);
```
Comment (daprahamian) [inline lib/options_validator.js]: I'm a little scared of this. Is it not possible for a dynamic default to be null b/c the inherited value is null?

Comment (mbroadst) [reply lib/options_validator.js]: Yeah, I meant for this to be a warning not an error. @daprahamian what is an example case where a default is `null`?

Comment (kvwalker) [reply lib/options_validator.js]: The `countDocuments should correctly call the given callback` test in `collection_test` currently warns that `serializeFunctions` is null. Based on the tests, it seems that there are many places where the bson options are not set on `this.s.`.

Comment (daprahamian) [reply lib/options_validator.js]: It seems like a lot of options can be:

1. If the user provided them for this call, use them,
2. Else if they set them on the mongoClient, use them
3. Else consider the option non-existent, and use a static default

Comment (mbroadst) [reply lib/options_validator.js]: @kvwalker let's make a ticket for the follow-on epic, remove the code check for whether the default is null/undefined, and unblock getting the rest of this PR in now.

Comment (kvwalker) [reply lib/options_validator.js]: Okay, made NODE-1748 and removed the code.

**Pull request merged**